### DISCUSSION TITLE
constraint in create table pass through reference table name validation

### DISCUSSION THREAD

**Discussion by tomocrafter** [sqldef/sqldef#617] [Feature requests]
## --export output
None

## Input SQL
Run following test in `cmd/psqldef/tests.yml`

```yaml
ConstraintInCreateTableBeforeCreateReferenceTable:
  desired: |
    CREATE TABLE posts (
      content text,
      user_id bigint,
      CONSTRAINT posts_ibfk_1 FOREIGN KEY (user_id) REFERENCES users (id) ON UPDATE CASCADE
    );
    CREATE TABLE users (id BIGINT PRIMARY KEY);
  output: |
    CREATE TABLE posts (
      content text,
      user_id bigint,
      CONSTRAINT posts_ibfk_1 FOREIGN KEY (user_id) REFERENCES users (id) ON UPDATE CASCADE
    );
    CREATE TABLE users (id BIGINT PRIMARY KEY);
```

## Current output
fails with
```
testutils.go:110: pq: relation "users" does not exist
```
instead of runtime

## Expected output
```
testutils.go:101: ALTER TABLE is performed before create table 'public.users': 'ALTER TABLE ONLY public.posts ADD CONSTRAINT posts_ibfk_1 FOREIGN KEY (user_id) REFERENCES users (id) ON UPDATE CASCADE'
```

I know that if runtime error occurs then migration transaction will be rollback'ed. so It does not affect current database.
So I wanted to make sure this is intended or not (since some of operations like `CREATE POLICY` validates table to exists).

Personally, I want it to fails at validation because dry-run doesn't fail although actual run fails.

**Comment by k0kubun** [comment]
> Personally, I want it to fails at validation because dry-run doesn't fail although actual run fails.

While we currently don't check semantic errors at a sqldef level, we're already simulating an in-progress schema for calculating the diff. I think it's not too hard (and it's nice) to validate the input while doing so. So I'd accept that feature. Would you be interested in filing a pull request to support it?

**Comment by gfx** [comment]
The recent sqldef orders output DDLs by its dependencies, so now it just works. Closing.
